### PR TITLE
[release] merging release-2022-01-14 into master

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @GuillaumeGomez @sdroege @bilelmoussaoui

### CODE REVIEW

**Comment by pbor** [comment]
Note: previous try already had a bunch of useful comments: https://github.com/gtk-rs/gtk-rs.github.io/pull/264

**Comment by bilelmoussaoui** [comment]
For gtk4-rs:
- The most important feature that landed on this release is a Rust BuilderScope implementation thanks to one of the awesome contributions of Jason Francis. This will allow applications to set callbacks/functions names in the UI file and define those in your Rust code. This only works with the CompositeTemplate macro.
- Trust upstream return nullability: in other words, gtk4 crates trust the nullable annotations provided in the GIR file which replaces a bunch of returned Option and returns only T nowadays. Please open an issue if you find any regressions
-  GTK 4.6 support, use v4_6 feature to unlock the new APIs
- Functions that return a glib::Value have now a corresponding function that gets you inner value T directly. Note those will panic if you provide the wrong type.
- GtkExpression helpers to make their usage nicer from Rust.
- Windows MSVC CI
- A bunch of fixed issues, documentation improvements and new book chapters.


other crates:
Other than the classic gtk-rs crates, we have also worked on ensuring the ones in https://gitlab.gnome.org/World/Rust received an update:
- libadwaita-rs: release 0.1 following up the release of the first stable version 1.0
- sourceview5-rs: updated to 0.4, with subclassing support
- libshumate-rs: released an alpha for the GTK4 Map widgetery library
- sourceview4-rs/libhandy-rs/libdazzle-rs/gspell-rs
- flatpak-rs: first release of the libflatpak bindings

**Comment by AaronErhardt** [comment]
Highlights in Relm4 v0.4:

### General
+ A component type called micro components was introduced to make applications more flexible at runtime
+ A type-safe API for actions

### Macros
+ A macro for creating menus
+ You can now use the `view` and the `menu` macro independently from Relm4 to build your UI in pure gtk4-rs applications!
+ New variants of the widget macro were ported to support Relm4's factories and micro components

---

Also, great work everyone! gtk4-rs is looking better than ever :+1:

**Comment by sdroege** [comment]
@bilelmoussaoui Can you add some links to the docs, and if anything would be interesting, some code snippets?

File: _posts/2022-01-14-new-release.md [hunk 1]
```diff
@@ -3,558 +3,574 @@ layout: post
...
+
+   let label_expression = button.property_expression("label");
+   ```
+ - A bunch of fixed issues, documentation improvements and new book chapters!
+
+## Other crates
+
+Other than the classic [gtk-rs][gtk-rs] crates, we have also worked on ensuring the ones in https://gitlab.gnome.org/World/Rust received an update:
+
+ - [libadwaita-rs](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/): release 0.1 following up the release of the first stable version 1.0
```
Comment (sdroege) [inline _posts/2022-01-14-new-release.md:214]: You can add gstreamer-rs here with the 0.18 release: https://gstreamer.freedesktop.org/news/#2022-01-16T11:00:00Z

Comment (GuillaumeGomez) [reply _posts/2022-01-14-new-release.md:214]: Done!

File: _posts/2022-01-14-new-release.md [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+The minimum supported versions of the underlying C libraries have not changed compared to last release but support for newer versions was added:
+
+ * GLib/GObject/GIO: minimum 2.48, newest 2.72
+ * Pango/PangoCairo: minimum 1.38, newest 1.52
+ * gdk-pixbuf: minimum 2.32, newest 2.42
+ * cairo: minimum 1.14, newest 1.16
+ * graphene: minimum 1.10
+ * ATK: minimum 2.18, newest 2.38
+ * GTK3: minimum 3.18, newest, 3.24.9
+ * GTK4: minimum 4.0, newest 4.6
```
Comment (sdroege) [inline _posts/2022-01-14-new-release.md:27]: Hm title says latest, text says newest :) Make it consistent, which one doesn't matter

File: _posts/2022-01-14-new-release.md [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+### GLib string builder
+
+`glib::String` was renamed to [`glib::GStringBuilder`](https://gtk-rs.org/gtk-rs-core/git/docs/glib/struct.GStringBuilder.html) to avoid confusion with `glib::GString` and to make it clearer that this is meant for building `NUL`-terminated C strings.
+
+In addition the API was simplified considerably:
+ * All APIs that allowed to create invalid UTF-8 were removed, leaving only safe APIs to build UTF-8 strings.
+ * `GStringBuilder` directly dereferences into `&str`, so all `&str` API can be used on it.
+ * `GStringBuilder` implements the `std::fmt::Write` trait, which allows making use of various `std` Rust API. For example `write!(&mut builder, "some format string {}", some_value)` can be used directly to write formatted strings into the string builder.
+
+### Multiple improvements on `ObjectExt`
```
Comment (sdroege) [inline _posts/2022-01-14-new-release.md:87]:
```suggestion
### Multiple improvements on `glib::ObjectExt`
```

File: _posts/2022-01-14-new-release.md [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+obj.connect_closure(
+    "notify", false,
+    glib::closure_local!(|_obj: glib::Object, pspec: glib::ParamSpec| {
+        println!("property notify: {}", pspec.name());
+    }));
+```
+
+The macros support the same syntax as the [`glib::clone`](https://gtk-rs.org/gtk-rs-core/stable/0.15/docs/glib/macro.clone.html) macro for creating strong/weak references, plus a [watch](https://gtk-rs.org/gtk-rs-core/stable/0.15/docs/glib/macro.closure.html#object-watching) feature that automatically disconnects the closure once the watched object disappears.
+
+### Widgets subclassing improvements
```
Comment (sdroege) [inline _posts/2022-01-14-new-release.md:161]:
```suggestion
### Object subclassing improvements
```

**Comment by sdroege** [comment]
All looking good to me otherwise, thanks for collecting!

File: _posts/2022-01-14-new-release.md [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+// [...]
+let data = surface.take_data().expect("Not the unique owner of the surface");
+// Send `data` to another thread or otherwise treat it as a `&mut [u8]`.
+// [...]
+let surface = data.into_inner();
+// Use the `surface` with Cairo APIs again.
+```
+
+This solves one of the common problems when using `Cairo` in multi-threaded applications.
+
```
Comment (sdroege) [inline _posts/2022-01-14-new-release.md:181]: ### GIO bindings improvements

The gio bindings saw many additions and improvements. Among other things it is now possible to implement the [`gio::Initable`](https://gtk-rs.org/gtk-rs-core/stable/0.15/docs/gio/struct.Initable.html) interface, the [`gio::IOExtention`](https://gtk-rs.org/gtk-rs-core/stable/0.15/docs/gio/struct.IOExtension.html) API is now available, and the [`gio::Task`](https://gtk-rs.org/gtk-rs-core/stable/0.15/docs/gio/struct.Task.html) bindings have been reworked to address soundness issues. More improvements to the `gio::Task` bindings are necessary for them to become nice to use from Rust.

File: _posts/2022-01-14-new-release.md [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+### Gdk3-wayland
+
+In this release, we generated the bindings for `gdk3-wayland`. You can now use the brand new [gdkwayland](https://crates.io/crates/gdkwayland) crate!
```
Comment (sdroege) [inline _posts/2022-01-14-new-release.md:184]: It's called `gdkwayland`, not `gdk3-wayland`

**Pull request merged**